unic/utils/src/codepoints.rs

```diff
@@ -29,19 +29,194 @@ use std::ops::Range;
 /// Reference: <http://unicode.org/glossary/#code_point>
 pub const CODEPOINTS_RANGE: Range<u32> = 0x0..(0x10_FFFF + 1);
 
-/// Range of Unicde Scalar Values.
+/// Range of Surrogate Code Points.
 ///
-/// Reference: <http://unicode.org/glossary/#unicode_scalar_value>
-const SCALAR_VALUE_RANGE_1: Range<u32> = 0x0..0xD800;
-const SCALAR_VALUE_RANGE_2: Range<u32> = (0xDFFF + 1)..(0x10_FFFF + 1);
+/// Reference: <http://unicode.org/glossary/#surrogate_code_point>
+pub const SURROGATE_RANGE: Range<u32> = 0xD800..(0xDFFF + 1);
 
 /// Check a code-point against `SURROGATE_CODEPOINTS_RANGE`.
 #[inline]
 pub fn iter_all_chars() -> Box<Iterator<Item = char>> {
-    Box::new(
-        // TODO: maybe use char::from_u32_unchecked()
-        SCALAR_VALUE_RANGE_1
-            .chain(SCALAR_VALUE_RANGE_2)
-            .filter_map(from_u32),
-    )
+    Box::new(CharIterator::new())
+}
+
+#[derive(Copy, Clone, Debug, Ord, PartialOrd, Eq, PartialEq)]
+/// An iterator over all `char`.
+pub struct CharIterator {
+    /// The codepoint of the smallest uniterated codepoint.
+    ///
+    /// If greater than or equal to `end`, iteration is finished.
+    ///
+    /// # Safety
+    ///
+    /// Must be a valid, non-surrogate codepoint.
+    current: u32,
+
+    /// The codepoint one greater than the largest uniterated codepoint.
+    ///
+    /// If less than or equal to `current`, iteration is finished.
+    ///
+    /// # Safety
+    ///
+    /// Must be a valid, non-surrogate codepoint.
```

Comment:
Yes, actually. `end` is one past the last uniterated codepoint. If we are to keep this I guess the correct safety requirement would be _Must be a valid, non-surrogate codepoint, or one greater than `char::MAX`_ for `end` and _Must not be a surrogate codepoint_ for `start`.